src/ray/gcs/pubsub/gcs_pub_sub.cc

```diff
@@ -44,52 +44,113 @@ Status GcsPubSub::SubscribeAll(const std::string &channel, const Callback &subsc
+        ray::gcs::RedisCallbackManager::instance().RemoveCallback(
+            channel->second.callback_index);
+        channel->second.callback_index = -1;
+        channel->second.pending_reply = false;
+
+        if (channel->second.command_queue.empty()) {
+          // We are unsubscribed and there are no more commands to process.
+          // Delete the channel.
+          channels_.erase(channel);
+        } else {
+          // Process the next item in the queue.
+          RAY_CHECK(channel->second.command_queue.front().is_subscribe);
+          RAY_CHECK_OK(ExecuteCommandIfPossible(channel_key, channel->second));
+        }
       } else if (reply->IsSubscribeCallback()) {
-        if (done) {
-          done(Status::OK());
+        {
+          // Set the callback index.
+          absl::MutexLock lock(&mutex_);
+          auto channel = channels_.find(channel_key);
+          RAY_CHECK(channel != channels_.end());
+          channel->second.callback_index = callback_index;
+          channel->second.pending_reply = false;
+          // Process the next item in the queue, if any.
+          if (!channel->second.command_queue.empty()) {
+            RAY_CHECK(!channel->second.command_queue.front().is_subscribe);
+            RAY_CHECK_OK(ExecuteCommandIfPossible(channel_key, channel->second));
+          }
+        }
+
+        if (command_done_callback) {
+          command_done_callback(Status::OK());
         }
       } else {
         const auto reply_data = reply->ReadAsPubsubData();
         if (!reply_data.empty()) {
           rpc::PubSubMessage message;
           message.ParseFromString(reply_data);
-          subscribe(message.id(), message.data());
+          command_subscribe_callback(message.id(), message.data());
         }
       }
-    }
-  };
-
-  int64_t out_callback_index;
-  auto status = redis_client_->GetPrimaryContext()->PSubscribeAsync(pattern, callback,
-                                                                    &out_callback_index);
-  if (id) {
-    absl::MutexLock lock(&mutex_);
-    // If the same pattern has been subscribed more than once, the last subscription takes
-    // effect.
-    subscribe_callback_index_[pattern] = out_callback_index;
+    };
+    status = redis_client_->GetPrimaryContext()->PSubscribeAsync(channel_key, callback,
+                                                                 callback_index);
+    channel.pending_reply = true;
+    channel.command_queue.pop_front();
+  } else if (!command.is_subscribe && channel.callback_index != -1) {
+    status = redis_client_->GetPrimaryContext()->PUnsubscribeAsync(channel_key);
+    channel.pending_reply = true;
+    channel.command_queue.pop_front();
+  } else if (!channel.pending_reply) {
+    // There is no in-flight command, but none of the queued commands are
+    // runnable. The caller must have sent a command out-of-order.
+    RAY_LOG(FATAL) << "Caller attempted a duplicate subscribe or unsubscribe to channel "
```

Comment:
When gcs server restarts, gcs client should re-subscribe. Now we must unsubscribe first if we want re-subscribe, how about add some comments to Subscribe API ?

Reply:
I don't think this is really necessary yet since the GCS is not fault-tolerant in any meaningful way right now.

Reply:
We have add test_gcs_fault_tolerance.py, which will test restart gcs server.

Reply:
Does that test require pubsub? If not, I'll just leave a TODO here that we should retry the last subscribe command.

Reply:
That test require pubsub.